src/sciline/container.py

```diff
@@ -48,47 +37,90 @@ def wrapper(*args, **kwargs):
-    # as lazy. We therefore use singleton-scope (to ensure Dask will recognize the
-    # task as the same object) and also wrap the function in dask.delayed.
-    scope = injector.singleton
-    func = _delayed(func)
-    tps = typing.get_type_hints(func)
-
-    def bind(binder: injector.Binder):
-        binder.bind(tps['return'], injector.inject(func), scope=scope)
-
-    return bind
-
-
-def make_container(funcs: List[Callable], /, *, lazy: bool = False) -> Container:
-    """
-    Create a :py:class:`Container` from a list of functions.
-
-    Parameters
-    ----------
-    funcs:
-        List of functions to be injected. Must be annotated with type hints.
-    lazy:
-        If True, the functions are wrapped in :py:func:`dask.delayed` before
-        being injected. This allows to build a dask graph from the container.
-    """
-    # Note that we disable auto_bind, to ensure we do not accidentally bind to
-    # some default values. Everything must be explicit.
-    return Container(
-        injector.Injector([_injectable(f) for f in funcs], auto_bind=False),
-        lazy=lazy,
-    )
+T = TypeVar('T')
+
+
+class UnsatisfiedRequirement(Exception):
+    pass
+
+
+class Container:
+    def __init__(self, funcs: List[Callable], /, *, lazy: bool = False):
+        """
+        Create a :py:class:`Container` from a list of functions.
+
+        Parameters
+        ----------
+        funcs:
+            List of functions to be injected. Must be annotated with type hints.
+        lazy:
+            If True, the functions are wrapped in :py:func:`dask.delayed` before
+            being injected. This allows to build a dask graph from the container.
+        """
+        self._providers: Dict[type, Callable] = {}
+        self._lazy: bool = lazy
+        self._cache: Dict[type, Any] = {}
+        for func in funcs:
+            self.insert(func)
+
+    def insert(self, provider: Callable) -> None:
+        key = get_type_hints(provider)['return']
+        if (origin := get_origin(key)) is not None:
+            args = get_args(key)
+            if len(args) != 1 and any(isinstance(arg, TypeVar) for arg in args):
+                raise ValueError(f'Cannot handle {key} with more than 1 argument')
+            key = origin if isinstance(args[0], TypeVar) else key
+        if key in self._providers:
+            raise ValueError(f'Provider for {key} already exists')
+        self._providers[key] = _delayed(provider)
+
+    Return = typing.TypeVar("Return")
+
+    def call(self, func: Callable[..., Return], bound: Optional[Any] = None) -> Return:
+        tps = get_type_hints(func)
+        del tps['return']
+        args: Dict[str, Any] = {}
+        for name, tp in tps.items():
+            if isinstance(tp, TypeVar):
+                tp = tp if bound is None else bound
+            elif (origin := get_origin(tp)) is not None:
+                if isinstance(get_args(tp)[0], TypeVar):
+                    tp = origin[bound]
```

Comment:
Shouldn't it also check if the number of ``get_args`` is exactly 1 here as well?

Reply:
I had assumed that the check in `insert` takes care of that, but there is indeed some inconsistency here that I need to fix.

Reply:
If all domain specific generic types will be sub class of `sl.Scope`,
we can use `isinstance(origin, sl.Scope)` instead of checking the length of `get_args` perhaps...?

But it may not work either since a subclass can have more type arguments than the parent class unless `sl.Scope` forces sub classes not to have more than the same number of type arguments via `__init_subclass__`....

Reply:
I am in the process of managing to make an improvement to the mechanism, by handling specific `TypeVar` instances as distinct when binding params of generics. This will ensure that functions provided by the user are not unintentionally bound together, and will furthermore allow for handling generics with multiple params (or multiple generics with distinct params). Will provide an update later.